### PR TITLE
Fix for API bug

### PR DESCRIPTION
**Pull request by Madu86**
The changes enclosed in this PR should allow MM codes to pass different number of point charges and coordinates during QM/MM MD. Specifically, I have made the following changes. 

1. Removed nxt_charges from `setQuickJob` and introduced them in `getQuickEnergy` and `getQuickEnergyGradients`  ( see for eg: https://github.com/Madu86/QUICK/blob/bugfix171/src/quick_api_test.f90#L122-L126)
2. Updated API test to feed different number of point charges during API test (https://github.com/Madu86/QUICK/blob/bugfix171/src/quick_api_test.f90#L110-L116). 
3. Isolated memory allocation and deallocation of point charge information arrays  in quick_molspec and quick_calculated modules.

Please check and let me know if you have suggestions.

### CODE REVIEW

**Review by agoetz — COMMENTED** [review]
@madu before changing the test, did you confirm that you reproduce the old api test when keeping the point charge number fixed?

This all looks like it should work as intended but I have two questions:

1. Why do we have duplicate arrays in Quick to store point charges and point charge gradients? E.g. we have `quick_api_type%ptchrg_grad` and `quick_qm_struct_type%ptchg_grad`. Should this be restructured later?

2. It seems complicated and error prone to pass an `api` flag in several places. Would it not make more sense to keep everything local to the API module and use a `reallocate` function instead?

**Comment by agoetz** [comment]
A reallocate function could be implemented in several ways. For instance always deallocating and reallocating:
```
if (nxt_ptchg > 0) then
  if (allocated(ptchg_crd)) deallocate(ptchg_crd)
  allocate(ptchg_crd(4, nxt_ptchg)
end if
```
or reallocate only if n has changed, either saving n from the previous step via save attribute, or explicitly checking dimensions:
```
if (nxt_ptchg > 0) then
  n = 0
  if (allocated(ptchg_crd)) n = size(ptchg_crd, 2)
  if (n /= nxt_ptchg) then
    deallocate(ptchg_crd)
    allocate(ptch_crd(4,nxt_ptchg))
  end if
end if
```

**Comment by Madu86** [comment]
> @Madu before changing the test, did you confirm that you reproduce the old api test when keeping the point charge number fixed?

Yes. If you comment this line: https://github.com/Madu86/QUICK/blob/bugfix171/src/quick_api_test.f90#L110, compile the code, you would get exactly the same (see this comparison: https://pastebin.com/pgGpi53B).

> Why do we have duplicate arrays in Quick to store point charges and point charge gradients? E.g. we have quick_api_type%ptchrg_grad and quick_qm_struct_type%ptchg_grad. Should this be restructured later?

You raised a good point. This is because I didn't want to change the base code every time we change our minds about API. :D At the time I was writing API, we didn't have a clear picture about 1) what is the best way to pass data from QUICK to AMBER (having an API with a single struct and use the struct in sander f90 file or pass arrays through subroutine call), 2) how we are going to do the MPI integration. To be on the safe side, I added a struct into API, stored all data in there. Of course, now we are done with the integration and its good to remove large arrays from this struct. But I would like to do this cleanup in a separate PR.

**Comment by Madu86** [comment]
> It seems complicated and error prone to pass an api flag in several places. Would it not make more sense to keep everything local to the API module and use a reallocate function instead?

I have eliminated them.

File: src/amber_interface.f90 [hunk 1]
```diff
@@ -159,7 +159,7 @@ subroutine qm2_quick_energy(escf,scf_mchg)
     !------------------END MPI/ALL NODES---------------------------------    
 
     !allocate essential variables
-    if (quick_first_call) call alloc(quick_molspec)
+    if (quick_first_call) call alloc(quick_molspec,.true.)
```
[the remark this answers is not in the document]
Comment (Madu86) [reply src/amber_interface.f90]: This is compiled but useless. I thought of removing this; but then there is CMake.. :D

File: src/modules/quick_api_module.f90 [hunk 2]
```diff
@@ -338,46 +328,108 @@ subroutine get_atom_types(natoms, atomic_numbers, natm_type, atm_type_id, ierr)
...
-  if(quick_api%nxt_ptchg>0) then
+  quick_api%nxt_ptchg = nxt_ptchg
+
+  ! set number of external atoms in quick_molspec
+  quick_molspec%nextatom  = quick_api%nxt_ptchg
+
+  if(quick_api%nxt_ptchg .gt. 0) then
+    call allocate_point_charge(.true., ierr)
     quick_api%ptchg_crd = ptchg_crd
     quick_api%ptchg_grad = ptchg_grad
```
[the remark this answers is not in the document]
Comment (Madu86) [reply src/modules/quick_api_module.f90]: You are right. I have fixed this.

**Comment by Madu86** [comment]
> A reallocate function could be implemented in several ways. For instance always deallocating and reallocating:
> 
> ```
> if (nxt_ptchg > 0) then
>   if (allocated(ptchg_crd)) deallocate(ptchg_crd)
>   allocate(ptchg_crd(4, nxt_ptchg)
> end if
> ```
> 
> or reallocate only if n has changed, either saving n from the previous step via save attribute, or explicitly checking dimensions:
> 
> ```
> if (nxt_ptchg > 0) then
>   n = 0
>   if (allocated(ptchg_crd)) n = size(ptchg_crd, 2)
>   if (n /= nxt_ptchg) then
>     deallocate(ptchg_crd)
>     allocate(ptch_crd(4,nxt_ptchg))
>   end if
> end if
> ```
> 
> And this could all be kept local to the API. The reallocate function could be implemented as part of the module where the corresponding arrays live.

This is smart. I have implemented the second form for reallocation of ptchg_crd and ptchg_grad arrays. See recent commits.

File: src/modules/quick_api_module.f90 [hunk 3]
```diff
@@ -338,48 +328,107 @@ subroutine get_atom_types(natoms, atomic_numbers, natm_type, atm_type_id, ierr)
...
+  
+  ! allocate memory only if external charges exist
+  if ( .not. allocated(quick_api%ptchg_crd)) allocate(quick_api%ptchg_crd(4,quick_api%nxt_ptchg), stat=ierr)
+
+  call realloc(quick_molspec,ierr)
+
+  if(isgrad) then
+    if ( .not. allocated(quick_api%ptchg_grad)) allocate(quick_api%ptchg_grad(3,quick_api%nxt_ptchg), stat=ierr)
+    quick_api%ptchg_grad =0.0d0
+    call realloc(quick_qm_struct,ierr)
```
Comment (agoetz) [inline src/modules/quick_api_module.f90:348]: It seems like you need that `call realloc()` only once if you put it outside of the `if(isgrad)` statement at the end of the routine.

Comment (agoetz) [reply src/modules/quick_api_module.f90:348]: Oh scratch that. Those are two different structs. I will delete my comment.

**Review by agoetz — APPROVED** [review]
Looks good!

**Pull request merged**